Training/interface/DataLoader_main.h

```diff
@@ -0,0 +1,931 @@
+
+      // file = OpenRootFile(file_name);
+      // tauTuple = std::make_shared<tau_tuple::TauTuple>(file.get(), true);
+
+      std::cout << "Number of files to process: " << input_files.size() << std::endl;
+      tauTuple = std::make_shared<TauTuple>("taus", input_files, true);
+      end_dataset = std::min(setup::end_dataset, tauTuple->GetEntries());
+
+      // histogram to calculate weights
+      // TH1::AddDirectory(kFALSE); // sets a global switch disabling the referencing
+      TFile* file_input = new TFile(setup::input_spectrum.c_str());
+      TFile* file_target = new TFile(setup::target_spectrum.c_str());
+      TH2D* target_hist = (TH2D*)file_target->Get("eta_pt_hist_tau");
+
+      for(int type = 0; type < 4; ++type) {
+        hist_weights.push_back(shared_ptr<TH2D>(new TH2D(("w_1_"+labels_names[type]).c_str(),
+                                                         ("w_1_"+labels_names[type]).c_str(),
+                                                         n_eta_bins, eta_min, eta_max,
+                                                         n_pt_bins, pt_min, pt_max)));
+        hist_weights[type]->SetDirectory(0);
+
+        TH2D* after_rebin_input_hist = new TH2D("input_hist", "input_hist",
+                                                n_eta_bins, eta_min, eta_max,
+                                                n_pt_bins, pt_min, pt_max);
+        TH2D* input_hist = (TH2D*)file_input->Get(("eta_pt_hist_"+labels_names[type]).c_str());
+
+        RebinAndFill(*hist_weights[type], *target_hist);
+        RebinAndFill(*after_rebin_input_hist, *input_hist);
+        hist_weights[type]->Divide(after_rebin_input_hist);
+        delete after_rebin_input_hist;
+      }
+
+      file_input->Close();
+      file_target->Close();
+    }
+
+    bool HasNext() {
+        return (current_entry + n_tau) < end_dataset;
+    }
+
+    size_t GetMaxBatchNumber () {
+        return (end_dataset - start_dataset) / n_tau;
+    }
+
+    std::shared_ptr<Data> LoadNext(){
+
+        if(!tauTuple)
+            throw std::runtime_error("DataLoader is not initialized.");
+
+        auto data = std::make_shared<Data>(n_tau, n_fe_tau,
+                                           n_inner_cells, n_outer_cells,
+                                           n_pf_el,
+                                           n_pf_mu,
+                                           n_pf_chHad,
+                                           n_pf_nHad,
+                                           n_pf_gamma,
+                                           n_electrons,
+                                           n_muons,
+                                           n_labels
+                                          );
+
+        const Long64_t end_entry = current_entry + n_tau;
+        size_t n_processed = 0, n_total = static_cast<size_t>(end_entry - start_dataset);
+        for(Long64_t tau_i = 0; tau_i < n_tau; ++current_entry) {
```

Comment:
`++current_entry, ++tau_i`, instead of having it on L336

Reply:
Unfortunately, here the idea was to increase tau_i counter only when we have needed type (tau.tauType < 0 || tau.tauType > 3) to select exactly n_tau candidates of required type, but if it is not readable at a glimpse I will think how to rearrange it in more elegant manner